### PR TITLE
check `type` of `path`

### PR DESCRIPTION
**Pull request by aaronspring**
otherwise uninformative error message: KeyError trying to assess first element of str instead of first element of list

### CODE REVIEW

File: intake_thredds/source.py [hunk 1]
```diff
@@ -69,6 +69,8 @@ def __init__(
         self.urlpath = url
         if 'simplecache::' in url:
             self.metadata.update({'fsspec_pre_url': 'simplecache::'})
+        assert isinstance(path, list), print('path must be list of str')
+        assert isinstance(list[0], str), print('path must be list of str')
```
Comment (aaronspring) [inline intake_thredds/source.py:73]:
```suggestion
        assert isinstance(path[0], str), print('path must be list of str')
```

File: intake_thredds/source.py [hunk 2]
```diff
@@ -69,6 +69,8 @@ def __init__(
         self.urlpath = url
         if 'simplecache::' in url:
             self.metadata.update({'fsspec_pre_url': 'simplecache::'})
+        assert isinstance(path, list), print('path must be list of str')
+        assert isinstance(path[0], str), print('path must be list of str')
```
Comment (andersy005) [inline intake_thredds/source.py:73]: The print function isn't necessary here

```suggestion
        assert isinstance(path, list), 'path must be list of str'
        assert isinstance(path[0], str), 'path must be list of str'
```

File: intake_thredds/source.py [hunk 3]
```diff
@@ -69,6 +69,8 @@ def __init__(
         self.urlpath = url
         if 'simplecache::' in url:
             self.metadata.update({'fsspec_pre_url': 'simplecache::'})
+        assert isinstance(path, list), 'path must be list of str'
+        assert isinstance(path[0], str), 'path must be list of str'
```
Comment (aaronspring) [inline intake_thredds/source.py:73]: alternatively we could convert a given str into a list
```suggestion
        if isinstance(path, str):
            path = list(path)
        assert isinstance(path, list), 'path must be list of str'
        assert isinstance(path[0], str), 'path must be list of str'
```

Comment (andersy005) [reply intake_thredds/source.py:73]: `path = list(path)` -> `path = [path]`

**Comment by dopplershift** [comment]
Peanut gallery here, but a more Pythonic way would be to `raise ValueError` for problems with input rather than an `assert`.

**Comment by aaronspring** [comment]
> Peanut gallery here, but a more Pythonic way would be to `raise ValueError` for problems with input rather than an `assert`.

thanks and implemented

File: intake_thredds/source.py [hunk 4]
```diff
@@ -69,6 +69,12 @@ def __init__(
         self.urlpath = url
         if 'simplecache::' in url:
             self.metadata.update({'fsspec_pre_url': 'simplecache::'})
+        if isinstance(path, str):
+            path = list(path)
```
Comment (andersy005) [inline intake_thredds/source.py:73]: Calling `list()` on a string will result in a list of characters.

```suggestion
            path =[path]
```

File: intake_thredds/source.py [hunk 5]
```diff
@@ -69,6 +69,12 @@ def __init__(
...
+        if not isinstance(path, list):
+            raise ValueError(f'path must be list of str, found {type(path)}')
+        if not isinstance(path[0], str):
+            raise ValueError(f'path must be list of str, found {type(path)}')
```
Comment (andersy005) [inline intake_thredds/source.py:77]: Do you mind adding one or two unit tests for this change?

File: intake_thredds/source.py (hunk 4 above)
Comment (aaronspring) [reply intake_thredds/source.py:73]: ah. now I get it. thanks

File: intake_thredds/source.py (hunk 5 above)
Comment (aaronspring) [reply intake_thredds/source.py:77]: added

Comment (andersy005) [reply intake_thredds/source.py:77]: Unfortunately, the unit tests appear not to cover these lines er [codecov](https://app.codecov.io/gh/NCAR/intake-thredds/blob/d1e4ca7906d9cbf75a4a7adfb018266f7e8c52d5/intake_thredds/source.py). I'm going to suggest changes shortly. Feel free to incorporate them

Screen Shot 2022-01-11 at 11 18 29 AM

File: tests/test_source.py [hunk 6]
```diff
@@ -42,6 +42,15 @@ def THREDDSMergedSource_cat_short(
     return cat
 
 
+@pytest.fixture(scope='module')
+@pytest.mark.parametrize(
+    'path', ['air.sig995.194*.nc', ['air.sig995.194*.nc'], pytest.param(1, marks=pytest.mark.xfail)]
+)
```
Comment (andersy005) [inline tests/test_source.py:48]: How about putting test for errors in a separate test?


```python
@pytest.mark.parametrize('path', [1, [1, 'air.sig995.194*.nc']])
def THREDDSMergedSource_path_error(THREDDSMergedSource_cat_short_url, path):
   with pytest.raises(ValueError):
      intake.open_thredds_merged(THREDDSMergedSource_cat_short_url, path)
```



```suggestion
@pytest.mark.parametrize(
    'path', ['air.sig995.194*.nc', ['air.sig995.194*.nc']]
)
```

File: intake_thredds/source.py [hunk 7]
```diff
@@ -69,6 +69,12 @@ def __init__(
         self.urlpath = url
         if 'simplecache::' in url:
             self.metadata.update({'fsspec_pre_url': 'simplecache::'})
+        if isinstance(path, str):
+            path = [path]
+        if not isinstance(path, list):
+            raise ValueError(f'path must be list of str, found {type(path)}')
+        if not isinstance(path[0], str):
```
Comment (andersy005) [inline intake_thredds/source.py:76]: I find this validation to be fragile. Shouldn't we be validating the entire list and not just the first element?

Comment (aaronspring) [reply intake_thredds/source.py]: implemented more robust

**Comment by aaronspring** [comment]
Thanks for the review and coding suggestions

**Pull request merged**